simplex/storage.go

```diff
@@ -0,0 +1,269 @@
+func newStorage(ctx context.Context, config *Config, qcDeserializer *QCDeserializer, blockTracker *blockTracker) (*Storage, error) {
+	lastAccepted, err := config.VM.LastAccepted(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	lastAcceptedBlock, err := config.VM.GetBlock(ctx, lastAccepted)
+	if err != nil {
+		return nil, err
+	}
+
+	genesisBlock, err := getGenesisBlock(ctx, config, blockTracker)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &Storage{
+		db:           config.DB,
+		genesisBlock: genesisBlock,
+		vm:           config.VM,
+		deserializer: qcDeserializer,
+		blockTracker: blockTracker,
+		log:          config.Log,
+	}
+	s.height.Store(lastAcceptedBlock.Height() + 1)
+
+	return s, nil
+}
+
+func (s *Storage) Height() uint64 {
+	return s.height.Load()
+}
+
+// Retrieve returns the block and finalization at [seq].
+// If [seq] is not found, returns false.
+func (s *Storage) Retrieve(seq uint64) (simplex.VerifiedBlock, simplex.Finalization, bool) {
+	if seq == 0 {
+		return s.genesisBlock, simplex.Finalization{}, true
+	}
+
+	block, err := getBlockAtHeight(context.TODO(), s.vm, seq)
+	if err != nil {
+		if err == database.ErrNotFound {
+			s.log.Error("block not found for sequence", zap.Uint64("seq", seq), zap.Error(err))
+		}
+		return nil, simplex.Finalization{}, false
+	}
+
+	finalization, found, err := s.retrieveFinalization(seq)
+	if !found || err != nil {
+		return nil, simplex.Finalization{}, false
+	}
+
+	vb := &Block{vmBlock: block, metadata: finalization.Finalization.ProtocolMetadata, blockTracker: s.blockTracker}
+	bytes, err := vb.Bytes()
+	if err != nil {
+		s.log.Error("failed to serialize block", zap.Error(err))
+		return nil, simplex.Finalization{}, false
+	}
+	vb.digest = computeDigest(bytes)
+
+	return vb, finalization, true
+}
+
+// Index indexes the finalization in the storage.
+// It stores the finalization bytes at the current height and increments the height.
+func (s *Storage) Index(ctx context.Context, block simplex.VerifiedBlock, finalization simplex.Finalization) error {
+	bh := block.BlockHeader()
+	if bh.Seq == 0 {
+		s.log.Warn("attempted to index genesis block, which should not be indexed")
+		return errGenesisIndexed
+	}
+
+	currentHeight := s.height.Load()
+	if currentHeight != bh.Seq {
+		return fmt.Errorf("%w: expected %d, got %d", errUnexpectedSeq, currentHeight, bh.Seq)
+	}
+
+	if s.blockTracker.lastIndexed != bh.Prev {
+		s.log.Warn("attempted to index block with mismatched previous digest",
+			zap.String("expected", s.blockTracker.lastIndexed.String()),
+			zap.String("got", bh.Prev.String()))
+
+		return fmt.Errorf("%w: expected %s, got %s", errMismatchedPrevDigest, s.blockTracker.lastIndexed, bh.Prev)
+	}
+
+	if !bytes.Equal(bh.Digest[:], finalization.Finalization.Digest[:]) {
+		return fmt.Errorf("%w: expected %d, got %d", errMismatchedDigest, bh.Digest, finalization.Finalization.Digest)
+	}
+
+	if finalization.QC == nil {
+		return errInvalidQC
+	}
```

Comment:
Would it be possible to make it more clear as to what checks are sanity checks vs required checks?

Perhaps we always `log.Error` on sanity checks?